subsys/bluetooth/mesh/adv.c

```diff
@@ -47,124 +46,148 @@ static K_FIFO_DEFINE(bt_mesh_adv_queue);
 		}
 	}
 }
 
-static void bt_mesh_adv_send_end(int err, struct bt_mesh_adv const *adv)
+static void bt_mesh_adv_send_end(int err, struct bt_mesh_adv_ctx const *ctx)
 {
-	if (adv->started && adv->cb && adv->cb->end) {
-		adv->cb->end(err, adv->cb_data);
+	if (ctx->started && ctx->cb && ctx->cb->end) {
+		ctx->cb->end(err, ctx->cb_data);
 	}
 }
 
-static void adv_buf_destroy(struct net_buf *buf)
+static struct bt_mesh_adv *adv_create_from_pool(struct k_mem_slab *buf_pool,
+						enum bt_mesh_adv_type type,
+						enum bt_mesh_adv_tag tag,
+						uint8_t xmit, k_timeout_t timeout)
 {
-	struct bt_mesh_adv adv = *BT_MESH_ADV(buf);
+	struct bt_mesh_adv_ctx *ctx;
+	struct bt_mesh_adv *adv;
+	int err;
 
-	net_buf_destroy(buf);
+	if (atomic_test_bit(bt_mesh.flags, BT_MESH_SUSPENDED)) {
+		LOG_WRN("Refusing to allocate buffer while suspended");
+		return NULL;
+	}
 
-	bt_mesh_adv_send_end(0, &adv);
-}
+	err = k_mem_slab_alloc(buf_pool, (void **)&adv, timeout);
+	if (err) {
+		return NULL;
+	}
 
-NET_BUF_POOL_DEFINE(adv_buf_pool, CONFIG_BT_MESH_ADV_BUF_COUNT,
-		    BT_MESH_ADV_DATA_SIZE, BT_MESH_ADV_USER_DATA_SIZE,
-		    adv_buf_destroy);
+	adv->__ref = 1;
 
-static struct bt_mesh_adv adv_local_pool[CONFIG_BT_MESH_ADV_BUF_COUNT];
+	net_buf_simple_init_with_data(&adv->b, adv->__bufs, BT_MESH_ADV_DATA_SIZE);
+	net_buf_simple_reset(&adv->b);
 
-#if defined(CONFIG_BT_MESH_RELAY)
-NET_BUF_POOL_DEFINE(relay_buf_pool, CONFIG_BT_MESH_RELAY_BUF_COUNT,
-		    BT_MESH_ADV_DATA_SIZE, BT_MESH_ADV_USER_DATA_SIZE,
-		    adv_buf_destroy);
+	ctx = &adv->ctx;
 
-static struct bt_mesh_adv adv_relay_pool[CONFIG_BT_MESH_RELAY_BUF_COUNT];
-#endif
+	(void)memset(ctx, 0, sizeof(*ctx));
 
-#if defined(CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE)
-NET_BUF_POOL_DEFINE(friend_buf_pool, CONFIG_BT_MESH_FRIEND_LPN_COUNT,
-		    BT_MESH_ADV_DATA_SIZE, BT_MESH_ADV_USER_DATA_SIZE,
-		    adv_buf_destroy);
+	ctx->type         = type;
+	ctx->tag          = tag;
+	ctx->xmit         = xmit;
 
-static struct bt_mesh_adv adv_friend_pool[CONFIG_BT_MESH_FRIEND_LPN_COUNT];
-#endif
+	return adv;
+}
 
-static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_pool,
-						    struct bt_mesh_adv *adv_pool,
-						    enum bt_mesh_adv_type type,
-						    enum bt_mesh_adv_tag tag,
-						    uint8_t xmit, k_timeout_t timeout)
+struct bt_mesh_adv *bt_mesh_adv_ref(struct bt_mesh_adv *adv)
 {
-	struct bt_mesh_adv *adv;
-	struct net_buf *buf;
+	__ASSERT_NO_MSG(adv->__ref < UINT8_MAX);
 
-	if (atomic_test_bit(bt_mesh.flags, BT_MESH_SUSPENDED)) {
-		LOG_WRN("Refusing to allocate buffer while suspended");
-		return NULL;
-	}
+	adv->__ref++;
 
-	buf = net_buf_alloc(buf_pool, timeout);
-	if (!buf) {
-		return NULL;
+	return adv;
+}
+
+void bt_mesh_adv_unref(struct bt_mesh_adv *adv)
+{
+	__ASSERT_NO_MSG(adv->__ref > 0);
+
+	if (--adv->__ref > 0) {
+		return;
 	}
 
-	adv = &adv_pool[net_buf_id(buf)];
-	BT_MESH_ADV(buf) = adv;
+	struct k_mem_slab *slab = NULL;
```

Comment:
```suggestion
	struct k_mem_slab *slab = &adv_buf_pool;
```

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/66130/files

Reply:
But `NULL` is also a value.

Reply:
Another suggestion:
```
	struct k_mem_slab *slab;
	struct bt_mesh_adv_ctx ctx = adv->ctx;

#if defined(CONFIG_BT_MESH_RELAY)
	if (adv->ctx.tag == BT_MESH_ADV_TAG_RELAY) {
		slab = &relay_buf_pool;
	} else
#endif

#if defined(CONFIG_BT_MESH_ADV_EXT_FRIEND_SEPARATE)
	if (adv->ctx.tag == BT_MESH_ADV_TAG_FRIEND) {
		slab = &friend_buf_pool;
	} else 
#endif

	{
		slab = &adv_buf_pool;
	}
```

Reply:
See this PR L153-L155

Reply:
Second option CI Failed, so first option.